libs/libc/sched/sched_gettid.c

```diff
@@ -0,0 +1,65 @@
+/****************************************************************************
+ * libs/libc/sched/sched_gettid.c
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+#include <nuttx/tls.h>
+
+#include <unistd.h>
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: nxsched_gettid
+ *
+ * Description:
+ *   Utilise thread local storage to get the thread ID of the currently
+ *   executing thread and cache the result for further queries. Caching this
+ *   result reduces system call loading when using CONFIG_BUILD_KERNEL.
+ *
+ * Input parameters:
+ *   None
+ *
+ * Returned Value:
+ *   On success, returns the thread ID of the calling process.
+ *
+ ****************************************************************************/
+
+#if !defined(CONFIG_BUILD_FLAT) && defined(CONFIG_SCHED_THREAD_LOCAL) && !defined(__KERNEL__)
+
+pid_t sched_gettid(void)
+{
+  FAR struct tls_info_s *tlsinfo = tls_get_info();
+
+  DEBUGASSERT(tlsinfo);
+
+  if (tlsinfo->tl_tid <= 0)
+    {
+        tlsinfo->tl_tid = gettid();
```

Comment:
should remove pid field from struct tcb_s and nxsched_gettid

Reply:
I must be missing something. The `tl_tid` is only read from userspace, and only built into the userspace libc. Even if a user application maliciously modified its value, it is never read from the kernel's context.

Again, I am newer to using NuttX, and there may be an interaction which I am overlooking.

Reply:
If we keep the origin kernel version isn't a problem.

Reply:
Sorry, can you explain a little more specifically what you mean.

Reply:
The better implementation is removing pid in struct tcb_s and let's both user and kernel use the same field(tl_tid) in tls_info_s. But using the same pid will sufer the secure hole I mention before.